From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Patrick McCarty <patrick.mccarty@intel.com>
Date: Wed, 22 Jul 2020 10:42:18 -0700
Subject: [PATCH] Enable static-pie linkage for binaries

Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>
---
 Makefile | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/Makefile b/Makefile
index 2dfffae..b1e3374 100644
--- a/Makefile
+++ b/Makefile
@@ -16,8 +16,8 @@ GO ?= go
 
 # test for go module support
 ifeq ($(shell go help mod >/dev/null 2>&1 && echo true), true)
-export GO_BUILD=GO111MODULE=on $(GO) build -mod=vendor
-export GO_TEST=GO111MODULE=on $(GO) test -mod=vendor
+export GO_BUILD=GO111MODULE=on $(GO) build -mod=vendor -buildmode=pie
+export GO_TEST=GO111MODULE=on $(GO) test -mod=vendor -buildmode=pie
 else
 export GO_BUILD=$(GO) build
 export GO_TEST=$(GO) test
@@ -33,7 +33,8 @@ BINDIR := /usr/local/bin
 
 VERSION := $(shell git describe --tags --dirty --always)
 VERSION := $(VERSION:v%=%)
-GO_LDFLAGS := -X $(PROJECT)/pkg/version.Version=$(VERSION)
+GO_LDFLAGS := -X $(PROJECT)/pkg/version.Version=$(VERSION) -extldflags "-static-pie"
+GO_TAGS := -tags netgo,osusergo
 
 BUILD_PATH := $(shell pwd)/build
 BUILD_BIN_PATH := $(BUILD_PATH)/bin
@@ -56,13 +57,13 @@ help:
 	@echo " * 'clean' - Clean artifacts."
 
 critest:
-	CGO_ENABLED=0 $(GO_TEST) -c -o $(CURDIR)/_output/critest$(BIN_EXT) \
-	     -ldflags '$(GO_LDFLAGS)' \
+	$(GO_TEST) -c -o $(CURDIR)/_output/critest$(BIN_EXT) \
+	     -ldflags '$(GO_LDFLAGS)' $(GO_TAGS) \
 	     $(PROJECT)/cmd/critest
 
 crictl:
-	CGO_ENABLED=0 $(GO_BUILD) -o $(CURDIR)/_output/crictl$(BIN_EXT) \
-		-ldflags '$(GO_LDFLAGS)' \
+	$(GO_BUILD) -o $(CURDIR)/_output/crictl$(BIN_EXT) \
+		-ldflags '$(GO_LDFLAGS)' $(GO_TAGS) \
 		$(PROJECT)/cmd/crictl
 
 clean:
